cpp/include/cudf/detail/aggregation/aggregation.hpp

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/detail/utilities/device_operators.cuh>
+#include <cudf/types.hpp>
+
+namespace cudf {
+namespace experimental {
+/**
+ * @brief Base class for specifying the desired aggregation in an
+ * `aggregation_request`.
+ *
+ * This type is meant to be opaque in the public interface.
+ *
+ * Other kinds of aggregations may derive from this class to encapsulate
+ * additional information needed to compute the aggregation.
+ */
+class aggregation {
+ public:
+  /**
+   * @brief Possible aggregation operations
+   */
+  enum Kind {
+    SUM,       ///< sum reduction
+    MIN,       ///< min reduction
+    MAX,       ///< max reduction
+    COUNT,     ///< count number of elements
+    MEAN,      ///< arithmetic mean reduction
+    MEDIAN,    ///< median reduction
+    QUANTILE,  ///< compute specified quantile(s)
+    ARGMAX,    ///< Index of max element
+    ARGMIN     ///< Index of min element
+  };
+
+  aggregation(aggregation::Kind a) : kind{a} {}
+  Kind kind;  ///< The aggregation to perform
+};
+namespace detail {
+/**
+ * @brief Derived class for specifying a quantile aggregation
+ */
+struct quantile_aggregation : aggregation {
+  quantile_aggregation(std::vector<double> const& q,
+                       experimental::interpolation i)
+      : aggregation{QUANTILE}, _quantiles{q}, _interpolation{i} {}
+  std::vector<double> _quantiles;              ///< Desired quantile(s)
+  experimental::interpolation _interpolation;  ///< Desired interpolation
```

Comment:
Data members in classes should be private. (Scott Meyers Effective C++ 3rd Ed. Item 22 Declare members private)